pkg/kubelet/util/csr/csr.go

```diff
@@ -50,30 +55,46 @@ func RequestNodeCertificate(client certificatesclient.CertificateSigningRequestI
 	if err != nil {
 		return nil, fmt.Errorf("unable to generate certificate request: %v", err)
 	}
-	return RequestCertificate(client, csrData, []certificates.KeyUsage{
+	hash := sha512.New512_256()
+	hash.Write(privateKeyData)
+	hash.Write([]byte(subject.CommonName))
+	for _, org := range subject.Organization {
+		hash.Write([]byte(org))
+	}
+	name := "node-csr-" + base64.StdEncoding.EncodeToString(hash.Sum(nil))
+	return requestCertificate(client, name, csrData, []certificates.KeyUsage{
 		certificates.UsageDigitalSignature,
 		certificates.UsageKeyEncipherment,
 		certificates.UsageClientAuth,
 	})
 }
 
-// RequestCertificate will create a certificate signing request using the PEM
+// requestCertificate will create a certificate signing request using the PEM
 // encoded CSR and send it to API server, then it will watch the object's
 // status, once approved by API server, it will return the API server's issued
 // certificate (pem-encoded). If there is any errors, or the watch timeouts, it
 // will return an error.
-func RequestCertificate(client certificatesclient.CertificateSigningRequestInterface, csrData []byte, usages []certificates.KeyUsage) (certData []byte, err error) {
+func requestCertificate(client certificatesclient.CertificateSigningRequestInterface, name string, csrData []byte, usages []certificates.KeyUsage) (certData []byte, err error) {
 	req, err := client.Create(&certificates.CertificateSigningRequest{
 		// Username, UID, Groups will be injected by API server.
-		TypeMeta:   metav1.TypeMeta{Kind: "CertificateSigningRequest"},
-		ObjectMeta: metav1.ObjectMeta{GenerateName: "csr-"},
-
+		TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
 		Spec: certificates.CertificateSigningRequestSpec{
 			Request: csrData,
 			Usages:  usages,
 		},
 	})
-	if err != nil {
+	switch {
+	case err == nil:
+	case errors.IsAlreadyExists(err):
+		glog.Infof("csr for this node already exists, reusing")
+		req, err = client.Get(name, metav1.GetOptions{})
```

Comment:
also need to make sure the existing csr object actually matches (subject, usages, and private key)